sdk/servicebus/service-bus/src/core/receiverHelper.ts

```diff
@@ -26,22 +50,25 @@ export class ReceiverHelper {
    * @returns true if credits were added, false if there is no current receiver instance
    * or `stopReceivingMessages` has been called.
    */
-  addCredit(credits: number): boolean {
-    const { receiver, logPrefix } = this._getCurrentReceiver();
+  addCredit(credits: number): void {
```

Comment:
> It checks and throws exceptions on much more granular conditions, particularly in addCredit

I understand that the changes here are related to the above statement made in the PR description.

There are 3 places where I see us adding credits
- when `subscribe()` is called
- when we are done awaiting on user provided message handler for a message 
- when we recover from a broken AMQP link or broken AMQP session

Do we expect the errors thrown from here to be surfaced to the user or is it more of an error for us to handle in the forever loop?

Reply:
For both 1 and 2 (normal subscribe and detach) they both run through _subscribeImpl, which will then put them into a retryForever loop. 

In retryForever the user will get the errors thrown if they exceed the normal retryOptions limit but it's merely a notification and not fatal. The only error that will stop this from happening is AbortError, which the user can do via actually aborting or by suspending the receiver.

For #2 - I've currently got that wrapped into a try/catch but it routes the error to _reportInternalError (which only forwards to processError if the set a special "SDK developers only" flag), rather than processError. I should flip that - then if we do get reports it'll show up via their normal processError callback and they won't have to turn on logging to see it.

Reply:
I see the initial subscribe() call and the detach cases entering into retryForever loop, but the loop does not apply to the sessions case where failure to add credit should result in the SessionLockLostError.

For case 2, if we are now moving to calling processError callback, would that also result in the error bubbling up to the retryForever loop?

Reply:
Sessions are interesting. I don't think I've actually altered the behavior there because the init() portion is done before we get to subscribe, unlike non-sessions. So by the time we reach the addCredit() call we've already validated that the link is open and exists in non-async calls so I don't even think addCredit can throw on anything that's not already checked. So there are no new error conditions here.

Case #2 is interesting, but I think we're good there. The processError isn't connected to a foreverLoop in that case (that's only used for the initial _subscribeImpl() call, and the call you're mentioning occurs inside of _onAmqpMessage). So they are two separate "threads" of execution , essentially. Calling processError there at least ensures that the error is reported, but our assumption (reasonably so) is that they'd only enter into that error because the receiver is reconnecting.

Reply:
Talked about this offline with @ramya-rao-a  and I get the concern now. For the record - the issue here is that in sessions if we (due to a bug) threw from addCredit() then the user might not realize that this is a fatal error even if an exception was thrown.

By convention we do throw SessionLockLost errors when fatal errors happen (we have an open issue to encode this convention into our session sample). I've updated the code to handle this and added in a test for it as well.

Also, for consistency this error is reported to processError, not just thrown (so either way we'd need a catch).